packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { CompletionItemKind, SnippetString } from 'vscode';
+import ProtocolCompletionItem from 'vscode-languageclient/lib/protocolCompletionItem';
+import { retrieveSObject, retrieveSObjects } from '../sfdx';
+
+import { Middleware } from 'vscode-languageclient';
+
+const EXPANDABLE_ITEM_PATTERN = /__([A-Z_]+)/;
+
+export const middleware: Middleware = {
+  // The SOQL LSP server may include special completion items as "placeholders" for
+  // the client to expand with information from the users' default Salesforce Org.
+  // We do that here as middleware, transforming the server response before passing
+  // it up to VSCode.
+
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const items = (await next(
+      document,
+      position,
+      context,
+      token
+    )) as ProtocolCompletionItem[];
+
+    return expandPlaceholders(await filterByContext(items));
+  }
+};
+
+interface SoqlItemContext {
+  sobjectName: string;
+  fieldName?: string;
+  notNillable?: boolean;
+}
+
+async function filterByContext(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const filteredItems: ProtocolCompletionItem[] = [];
+
+  for (const item of items) {
+    if (
+      !EXPANDABLE_ITEM_PATTERN.test(item.label) &&
+      item?.data?.soqlContext?.sobjectName
+    ) {
+      const objMetadata = await retrieveSObject(
+        item.data.soqlContext.sobjectName
+      );
+      const fieldMeta = objMetadata.fields.find(
+        field => field.name === item.data.soqlContext.fieldName
+      );
+      if (fieldMeta && !operatorsByType[fieldMeta.type].includes(item.label)) {
+        continue;
+      }
+    }
+
+    filteredItems.push(item);
+  }
+
+  return filteredItems;
+}
+
+async function expandPlaceholders(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const expandedItems = [...items];
+
+  for (const [index, item] of items.entries()) {
+    const parsedCommand = item.label.match(EXPANDABLE_ITEM_PATTERN);
+    if (parsedCommand) {
+      const commandName = parsedCommand[1];
+
+      const handler = expandFunctions[commandName];
+      if (!handler) {
+        console.log(`Unknown SOQL completion command ${commandName}!`);
+      }
+
+      expandedItems.splice(
+        index,
+        1,
+        ...(await handler(item?.data?.soqlContext))
+      );
+    }
+  }
+
+  return expandedItems;
+}
+
+const expandFunctions: {
+  [key: string]: (
+    soqlContext?: SoqlItemContext
+  ) => Promise<ProtocolCompletionItem[]>;
+} = {
+  SOBJECTS_PLACEHOLDER: async (
```

Comment:
expandFunctions is now doing lot more things that I originally thought it would do (which is just expand sobjects and its fields). 

I feel all this logic should move over to the language server instead of the client. I think we just go ahead and implement what we discussed in the last meeting to have the language server take in something like an sObjectProvider and move all the code here to the language server and keep the client simple.

Reply:
Also, SOQL completion in Apex would rely on the SOQL language server and we cannot have this logic duplicated over there as well in the Apex extension.

Reply:
Agreed.  We have to define how to do it though.... it's not trivial and we need to agree on an approach.

BTW: as an intermediate solution, some of this logic can be moved to the server-side. For example, for this new logic for type-aware fields, the server could send all completion items with extra metadata indicating for which types the item applies... so the client would only (and optionally) filter them based on the field type. Still not a solution, but wanted to mentioning it.